plugins/hls-class-plugin/src/Ide/Plugin/Class/CodeLens.hs

```diff
@@ -0,0 +1,133 @@
+{-# LANGUAGE GADTs           #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# OPTIONS_GHC -Wno-overlapping-patterns #-}
+
+module Ide.Plugin.Class.CodeLens where
+
+import           Control.Lens                    ((^.))
+import           Control.Monad.IO.Class          (liftIO)
+import           Data.Aeson
+import           Data.Maybe                      (mapMaybe)
+import qualified Data.Text                       as T
+import           Development.IDE
+import           Development.IDE.GHC.Compat
+import           Development.IDE.GHC.Compat.Util
+import           GHC.LanguageExtensions.Type
+import           Ide.Plugin.Class.Types
+import           Ide.Plugin.Class.Utils
+import           Ide.PluginUtils
+import           Ide.Types
+import           Language.LSP.Server             (sendRequest)
+import           Language.LSP.Types
+import qualified Language.LSP.Types.Lens         as J
+
+codeLens :: PluginMethodHandler IdeState TextDocumentCodeLens
+codeLens state plId CodeLensParams{..} = do
+    enabled <- enableTypeLens <$> getCompletionsConfig plId
+    if not enabled then pure $ pure $ List [] else pluginResponse $ do
+        nfp <- getNormalizedFilePath plId uri
+        tmr <- handleMaybeM "Unable to typecheak"
+            $ liftIO
+            $ runAction "classplugin.TypeCheck" state
+            $ use TypeCheck nfp
+
+        InstanceBindTypeSigsResult binds <-
+            handleMaybeM "Unable to get InstanceBindTypeSigsResult"
+            $ liftIO
+            $ runAction "classplugin.GetInstanceBindTypeSigs" state
+            $ use GetInstanceBindTypeSigs nfp
+
+        pragmaInsertion <- insertPragmaIfNotPresent state nfp InstanceSigs
+
+        let (hsGroup, _, _, _) = tmrRenamed tmr
+            tycls = hs_tyclds hsGroup
+            -- class instance decls
+            insts = mapMaybe (getClsInstD . unLoc) $ concatMap group_instds tycls
+            -- Declared instance methods without signatures
+            bindInfos = concatMap getBindSpanWithoutSig insts
```

Comment:
I don't normally suggest this, but perhaps some of these things would be nicer with list comprehensions:
```
bindInfos = [ bind | bind <- getBindSpanWithoutSig inst, Just inst <- getClsInstD $ unLoc inst_d, inst_d <- group_instds tycls ]
```